sympy/solvers/ode/single.py

```diff
@@ -0,0 +1,667 @@
+    >>> eq = f(x).diff(x, 2)
+
+    Now we solve this problem using the NthAlgebraic solver which is a
+    subclass of SingleODESolver:
+
+    >>> from sympy.solvers.ode.single import NthAlgebraic, SingleODEProblem
+    >>> problem = SingleODEProblem(eq, f(x), x)
+    >>> solver = NthAlgebraic(problem)
+    >>> solver.get_general_solution()
+    [Eq(f(x), _C*x + _C)]
+
+    The normal way to solve an ODE is to use dsolve (which would use
+    NthAlgebraic and other solvers internally). When using dsolve a number of
+    other things are done such as evaluating integrals, simplifying the
+    solution and renumbering the constants:
+
+    >>> from sympy import dsolve
+    >>> dsolve(eq, hint='nth_algebraic')
+    Eq(f(x), C1 + C2*x)
+    """
+
+    # Subclasses should store the hint name (the argument to dsolve) in this
+    # attribute
+    hint = None  # type: ClassVar[str]
+
+    # Subclasses should define this to indicate if they support an _Integral
+    # hint.
+    has_integral = None  # type: ClassVar[bool]
+
+    # The ODE to be solved
+    ode_problem = None  # type: SingleODEProblem
+
+    # Cache whether or not the equation has matched the method
+    _matched = None  # type: Optional[bool]
+
+    def __init__(self, ode_problem):
+        self.ode_problem = ode_problem
+
+    def matches(self) -> bool:
+        if self._matched is None:
+            self._matched = self._matches()
+        return self._matched
+
+    def get_general_solution(self, *, simplify: bool = True) -> List[Equality]:
+        if not self.matches():
+            msg = "%s solver can not solve:\n%s"
+            raise ODEMatchError(msg % (self.hint, self.ode_problem.eq))
+        return self._get_general_solution()
+
+    def _matches(self) -> bool:
+        msg = "Subclasses of SingleODESolver should implement matches."
+        raise NotImplementedError(msg)
+
+    def _get_general_solution(self, *, simplify: bool = True) -> List[Equality]:
+        msg = "Subclasses of SingleODESolver should implement get_general_solution."
+        raise NotImplementedError(msg)
+
+
+class SinglePatternODESolver(SingleODESolver):
+    '''Superclass for ODE solvers based on pattern matching'''
+
+    def wilds(self):
+        prob = self.ode_problem
+        f = prob.func.func
+        x = prob.sym
+        order = prob.order
+        return self._wilds(f, x, order)
+
+    def wilds_match(self):
+        match = self._wilds_match
+        return [match.get(w, S.Zero) for w in self.wilds()]
+
+    def _matches(self):
+        eq = self.ode_problem.eq_expanded
+        f = self.ode_problem.func.func
+        x = self.ode_problem.sym
+        order = self.ode_problem.order
+        df = f(x).diff(x)
+
+        if order != 1:
+            return False
+
+        eq = expand(eq / eq.coeff(df))
+        eq = eq.collect(f(x))
+
+        pattern = self._equation(f(x), x, 1)
+
+        self._wilds_match = match = eq.match(pattern)
+
+        return match is not None
```

Comment:
@oscarbenjamin, This method may require further modifications for cases like `Riccati_special_minus2` where after matching, it is checked if some of the coefficients are zero.
https://github.com/sympy/sympy/blob/35be3a042e38b1ef5de33120e38293540f40d5f5/sympy/solvers/ode/ode.py#L1091-L1099
Should we implement some other method for matching such ode or is there a workaround for this?

Reply:
Is it necessary for those to be non-zero or would the general solution apply equally well when those coefficients are zero?

Reply:
It is necessary for b i.e., `r[b2]` to be non-zero as the denominator term is `2*b*x` in the solution. But I think this can be resolved by using 
`b = Wild('b', exclude=[x, f(x), f(x).diff(x), 0])` in `_wilds` method. The other problem is checking the or condition (in line 1094 in above code segment) for `r[c2]` and `r[d2]`. Here, if both of them are zero, then the solution is not real.

Reply:
We also need to bear in mind the possibility that the coefficient might be a symbol that is not known to be zero. We can generally handle that sort of thing using a Piecewise in sympy e.g.:
```
In [282]: x = Symbol('x')                                                                                                                      

In [283]: Piecewise((1/x, Ne(x, 0)), (0, True))                                                                                                
Out[283]: 
⎧1           
⎪─  for x ≠ 0
⎨x           
⎪            
⎩0  otherwise

In [284]: Piecewise((1/x, Ne(x, 0)), (0, True)).subs(x, 2)                                                                                     
Out[284]: 1/2
```